File: libs/shared/npm/src/lib/local-nx-utils/corepack-detection.ts
import { readJsonFile } from '@nx-console/shared-file-system';
import { Logger } from '@nx-console/shared-utils';
import { join } from 'path';

/**
 * Detects if Corepack is being used by checking the packageManager field in package.json
 * @param workspacePath The workspace path
 * @param logger Optional logger
 * @returns The packageManager string (e.g., "yarn@4.7.0") or undefined if not using Corepack
 */
export async function detectCorepackPackageManager(
  workspacePath: string,
  logger?: Logger,
): Promise<string | undefined> {
  try {
    const packageJsonPath = join(workspacePath, 'package.json');
    const { json } = await readJsonFile(packageJsonPath);
    
    if (json.packageManager && typeof json.packageManager === 'string') {
      logger?.log(`Detected Corepack package manager: ${json.packageManager}`);
      return json.packageManager;
    }
    
    return undefined;
  } catch (e) {
    logger?.log(`Error detecting Corepack package manager: ${JSON.stringify(e)}`);
    return undefined;
  }
}

/**
 * Checks if Corepack should be used for the given workspace
 * @param workspacePath The workspace path
 * @param logger Optional logger
 * @returns true if Corepack should be used
 */
export async function shouldUseCorepack(
  workspacePath: string,
  logger?: Logger,
): Promise<boolean> {
  const packageManager = await detectCorepackPackageManager(workspacePath, logger);
  return packageManager !== undefined;
}

/**
 * Extracts the package manager name from a Corepack packageManager string
 * @param packageManagerString The packageManager string (e.g., "yarn@4.7.0")
 * @returns The package manager name (e.g., "yarn")
 */
export function extractPackageManagerName(packageManagerString: string): string {
  const match = packageManagerString.match(/^([^@]+)/);
  return match ? match[1] : packageManagerString;
}

File: apps/intellij/src/main/kotlin/dev/nx/console/utils/CorepackDetection.kt
package dev.nx.console.utils

import com.google.gson.JsonParser
import com.intellij.openapi.diagnostic.logger
import java.io.File
import java.nio.file.Files
import java.nio.file.Path

private val logger = logger<CorepackDetection>()

class CorepackDetection {
    companion object {
        /**
         * Detects if Corepack is being used by checking the packageManager field in package.json
         * @param workspacePath The workspace path
         * @return The packageManager string (e.g., "yarn@4.7.0") or null if not using Corepack
         */
        fun detectCorepackPackageManager(workspacePath: String): String? {
            try {
                val packageJsonPath = Path.of(workspacePath, "package.json")
                if (!Files.exists(packageJsonPath)) {
                    return null
                }
                
                val packageJsonContent = Files.readString(packageJsonPath)
                val jsonElement = JsonParser.parseString(packageJsonContent)
                
                if (jsonElement.isJsonObject) {
                    val jsonObject = jsonElement.asJsonObject
                    if (jsonObject.has("packageManager")) {
                        val packageManager = jsonObject.get("packageManager").asString
                        logger.info("Detected Corepack package manager: $packageManager")
                        return packageManager
                    }
                }
                
                return null
            } catch (e: Exception) {
                logger.warn("Error detecting Corepack package manager", e)
                return null
            }
        }
        
        /**
         * Checks if Corepack should be used for the given workspace
         * @param workspacePath The workspace path
         * @return true if Corepack should be used
         */
        fun shouldUseCorepack(workspacePath: String): Boolean {
            return detectCorepackPackageManager(workspacePath) != null
        }
        
        /**
         * Extracts the package manager name from a Corepack packageManager string
         * @param packageManagerString The packageManager string (e.g., "yarn@4.7.0")
         * @return The package manager name (e.g., "yarn")
         */
        fun extractPackageManagerName(packageManagerString: String): String {
            val atIndex = packageManagerString.indexOf('@')
            return if (atIndex > 0) {
                packageManagerString.substring(0, atIndex)
            } else {
                packageManagerString
            }
        }
    }
}